File: src/redstone.c
/* redstone.c - Redstone logic implementation
 *
 * Copyright (C) 2014 Ryan Mendivil <ryan@nullreff.net>
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "redstone.h"
#include "redpile.h"

#define MAX_POWER 15

#define LOCATION(NODE)  ((NODE)->location)
#define MATERIAL(NODE)  ((NODE)->type)
#define POWER(NODE)     FIELD_GET(NODE, 0)
#define DIRECTION(NODE) FIELD_GET(NODE, 1)
#define STATE(NODE)     FIELD_GET(NODE, 2)

#define LOWER_POWER(NODE,POWER) messages_power_check(data->in, LOCATION(NODE), POWER)
#define NODE_ADJACENT(NODE,DIR) world_get_adjacent_node(data->world, NODE, DIR)
#define MOVE_TO_NODE(NODE,DIR) NODE = NODE_ADJACENT(NODE, DIR)

#define SEND(NODE,CMD,VALUE,DELAY) queue_add(data->messages, CMD,            data->world->ticks + (DELAY), data->node, NODE, VALUE)
#define CMD_POWER(POWER)           queue_add(data->sets,     MESSAGE_POWER,  data->world->ticks,           data->node, data->node, POWER)
#define CMD_MOVE(DIR)              queue_add(data->sets,     MESSAGE_PUSH,   data->world->ticks,           data->node, data->node, DIR  )
#define CMD_REMOVE()               queue_add(data->sets,     MESSAGE_REMOVE, data->world->ticks,           data->node, data->node, 0    )

#define AIR        (data->world->types->data + 0)
#define INSULATOR  (data->world->types->data + 1)
#define WIRE       (data->world->types->data + 2)
#define CONDUCTOR  (data->world->types->data + 3)
#define TORCH      (data->world->types->data + 4)
#define PISTON     (data->world->types->data + 5)
#define REPEATER   (data->world->types->data + 6)
#define COMPARATOR (data->world->types->data + 7)
#define SWITCH     (data->world->types->data + 8)

TYPE_BEHAVIOR(PUSH_MOVE)
{
    Message* move_inst = messages_find_first(data->in, MESSAGE_PUSH | MESSAGE_PULL);
    if (move_inst != NULL)
    {
        CMD_MOVE(move_inst->message);
        return true;
    }
    return false;
}

TYPE_BEHAVIOR(PUSH_BREAK)
{
    Message* move_inst = messages_find_first(data->in, MESSAGE_PUSH);
    if (move_inst != NULL)
    {
        CMD_REMOVE();
        return true;
    }
    return false;
}

TYPE_BEHAVIOR(WIRE)
{
    Node* above = NODE_ADJACENT(data->node, UP);
    bool covered = above != NULL && MATERIAL(above) != AIR && MATERIAL(above) != EMPTY;

    int new_power = messages_max_power(data->in);
    CMD_POWER(new_power);

    if (new_power == 0)
        return true;

    int wire_power = new_power;
    if (wire_power > 0)
        wire_power--;

    // Adjacent blocks
    Direction directions[4] = {NORTH, SOUTH, EAST, WEST};
    for (int i = 0; i < 4; i++)
    {
        Direction dir = directions[i];
        // Directly adjacent
        Node* found_node = NODE_ADJACENT(data->node, dir);

        // Down one block
        if (MATERIAL(found_node) == AIR)
        {
            MOVE_TO_NODE(found_node, DOWN);
            if (MATERIAL(found_node) != WIRE)
                continue;
        }
        // Up one block
        else if (MATERIAL(found_node) == CONDUCTOR)
        {
            // Add charge to the block
            Node* right = NODE_ADJACENT(data->node, direction_left(dir));
            Node* left = NODE_ADJACENT(data->node, direction_right(dir));

            if (MATERIAL(right) != WIRE && MATERIAL(left) != WIRE &&
                LOWER_POWER(found_node, wire_power))
            {
                SEND(found_node, MESSAGE_POWER, wire_power, 0);
            }

            if (covered)
                continue;

            MOVE_TO_NODE(found_node, UP);
            if (MATERIAL(found_node) != WIRE)
                continue;
        }

        if (LOWER_POWER(found_node, wire_power))
            SEND(found_node, MESSAGE_POWER, wire_power, 0);
    }

    // Block below
    Node* down_node = NODE_ADJACENT(data->node, DOWN);
    if (MATERIAL(down_node) == CONDUCTOR && LOWER_POWER(down_node, new_power))
        SEND(down_node, MESSAGE_POWER, new_power, 0);

    return true;
}

TYPE_BEHAVIOR(CONDUCTOR)
{
    unsigned int power = messages_max_power(data->in);
    CMD_POWER(power);
    if (power == 0)
        return true;

    bool max_powered = power == MAX_POWER;

    for (int i = 0; i < DIRECTIONS_COUNT; i++)
    {
        Direction dir = (Direction)i;
        Node* found_node = NODE_ADJACENT(data->node, dir);
        if (MATERIAL(found_node) == CONDUCTOR)
            continue;

        if (!max_powered && MATERIAL(found_node) == WIRE)
            continue;

        if (LOWER_POWER(found_node, power))
            SEND(found_node, MESSAGE_POWER, power, 0);
    }

    return true;
}

TYPE_BEHAVIOR(TORCH)
{
    unsigned int new_power = 0;
    Location loc_behind = location_move(LOCATION(data->node), direction_invert(DIRECTION(data->node)), 1);
    for (int i = 0; i < data->in->size; i++)
    {
        Message* inst = data->in->data + i;

        // Power coming from behind
        if (LOCATION_EQUALS(inst->source.location, loc_behind))
            new_power = inst->message;
    }

    CMD_POWER(new_power);

    if (new_power > 0)
        return true;

    // Pass charge to any adjacent wires
    Direction behind = direction_invert(DIRECTION(data->node));
    Direction directions[5] = {NORTH, SOUTH, EAST, WEST, DOWN};
    for (int i = 0; i < 5; i++)
    {
        Direction dir = directions[i];
        if (dir == behind)
            continue;

        Node* found_node = NODE_ADJACENT(data->node, dir);
        SEND(found_node, MESSAGE_POWER, MAX_POWER, 1);
    }

    // Pass charge up to a conductor
    Node* up_node = NODE_ADJACENT(data->node, UP);
    if (MATERIAL(up_node) == CONDUCTOR)
        SEND(up_node, MESSAGE_POWER, MAX_POWER, 1);

    return true;
}

#define RETRACTED 0
#define RETRACTING 1
#define EXTENDED 2
#define EXTENDING 3
TYPE_BEHAVIOR(PISTON)
{
    Node* first = NODE_ADJACENT(data->node, DIRECTION(data->node));
    Node* second = NODE_ADJACENT(first, DIRECTION(data->node));

    unsigned int new_power = messages_max_power(data->in);

    unsigned int state;
    if (new_power == 0)
    {
        if (MATERIAL(first) == AIR && MATERIAL(second) != AIR && POWER(data->node) != 0)
            state = RETRACTING;
        else
            state = RETRACTED;
    }
    else
    {
        if (MATERIAL(second) == AIR && MATERIAL(first) != AIR && POWER(data->node) == 0)
            state = EXTENDING;
        else
            state = EXTENDED;
    }

    if (state == RETRACTED)
        return false;

    CMD_POWER(new_power);

    if (state == EXTENDING)
        SEND(first, MESSAGE_PUSH, DIRECTION(data->node), 1);
    else if (state == RETRACTING)
        SEND(second, MESSAGE_PULL, direction_invert(DIRECTION(data->node)), 1);

    return true;
}

TYPE_BEHAVIOR(REPEATER)
{
    bool side_powered = false;
    int new_power = 0;
    Location right  = location_move(LOCATION(data->node), direction_right(DIRECTION(data->node)), 1);
    Location left   = location_move(LOCATION(data->node), direction_left(DIRECTION(data->node)), 1);
    Location behind = location_move(LOCATION(data->node), direction_invert(DIRECTION(data->node)), 1);
    
    for (int i = 0; i < data->in->size; i++)
    {
        Message* inst = data->in->data + i;

        // Power coming from the side
        if (inst->source.type == REPEATER &&
            (LOCATION_EQUALS(inst->source.location, right) ||
            LOCATION_EQUALS(inst->source.location, left)))
        {
            side_powered = (inst->message > 0) || side_powered;
        }

        // Power coming from behind
        if (LOCATION_EQUALS(inst->source.location, behind))
            new_power = inst->message;
    }

    CMD_POWER(new_power);

    if (new_power == 0 || side_powered)
        return true;

    // Pass charge to the wire or conductor in front after a delay
    Node* found_node = NODE_ADJACENT(data->node, DIRECTION(data->node));
    SEND(found_node, MESSAGE_POWER, MAX_POWER, STATE(data->node) + 1);

    return true;
}

TYPE_BEHAVIOR(COMPARATOR)
{
    unsigned int side_power = 0;
    int new_power = 0;
    Location right  = location_move(LOCATION(data->node), direction_right(DIRECTION(data->node)), 1);
    Location left   = location_move(LOCATION(data->node), direction_left(DIRECTION(data->node)), 1);
    Location behind = location_move(LOCATION(data->node), direction_invert(DIRECTION(data->node)), 1);
    
    for (int i = 0; i < data->in->size; i++)
    {
        Message* inst = data->in->data + i;

        // Power coming from the side
        if ((LOCATION_EQUALS(inst->source.location, right) ||
             LOCATION_EQUALS(inst->source.location, left)) &&
            side_power < inst->message)
        {
            side_power = inst->message;
        }

        // Power coming from behind
        if (LOCATION_EQUALS(inst->source.location, behind))
            new_power = inst->message;
    }

    CMD_POWER(new_power);

    if (new_power == 0)
        return true;

    int change = new_power;

    // Subtraction mode
    if (STATE(data->node) > 0)
        change -= side_power;

    new_power = new_power > side_power ? change : 0;

    if (new_power == 0)
        return true;

    // Pass charge to the wire or conductor in front
    Node* found_node = NODE_ADJACENT(data->node, DIRECTION(data->node));
    SEND(found_node, MESSAGE_POWER, new_power, 1);

    return true;
}

TYPE_BEHAVIOR(SWITCH)
{
    if (STATE(data->node) == 0)
        return true;

    Direction behind = direction_invert(DIRECTION(data->node));
    for (Direction dir = (Direction)0; dir < DIRECTIONS_COUNT; dir++)
    {
        Node* found_node = NODE_ADJACENT(data->node, dir);
        if (MATERIAL(found_node) != CONDUCTOR || dir == behind)
            SEND(found_node, MESSAGE_POWER, MAX_POWER, 0);
    }

    return true;
}

File: conf/redstone.lua
-- redstone.lua - Redstone logic implementation
--
-- Copyright (C) 2014 Ryan Mendivil <ryan@nullreff.net>
-- All rights reserved.
--
-- Redistribution and use in source and binary forms, with or without
-- modification, are permitted provided that the following conditions are met:
--
--   * Redistributions of source code must retain the above copyright notice,
--     this list of conditions and the following disclaimer.
--   * Redistributions in binary form must reproduce the above copyright
--     notice, this list of conditions and the following disclaimer in the
--     documentation and/or other materials provided with the distribution.
--   * Neither the name of Redpile nor the names of its contributors may be
--     used to endorse or promote products derived from this software without
--     specific prior written permission.
--
-- THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
-- AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
-- IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
-- ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
-- LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
-- CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
-- SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
-- INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
-- CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
-- ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
-- POSSIBILITY OF SUCH DAMAGE.
--

MAX_POWER = 15
RETRACTED  = 0
RETRACTING = 1
EXTENDED   = 2
EXTENDING  = 3

function has_lower_power(node, messages, power)
    local received_power = messages:source(node.location)
    return received_power == nil or received_power.value < power
end

function value_or_zero(message)
    return message and message.value or 0
end

-- Behaviors are defined using the `redpile.behavior` function which takes:
--
-- NAME (String)
-- The name is what will be used to reference a behavior later when we attach
-- it to a type.
--
-- MESSAGES <Table>
-- Determines which message types this behavior listens for.  Message types are
-- specified when calling `node:send(message_type, ...)`.  Do not list a
-- message type more than once.
--
-- BEHAVIOR (Function)
-- This function takes a reference to the current node and a list of messages
-- received from other nodes.  It can then do whatever calculations it needs
-- to, send messages to other nodes and make modifications to the world.
-- However, be aware of two things:
--   1. Any modifications to the world will not occur until after the current
--   tick has completely finished processing all nodes.  If you need to
--   communicate a change with another node, send it a message.
--   2. This function may be called multiple times before, during, and after
--   the tick or may be cached and never run again.  Therefore it needs be
--   'pure' and not use any global state in the program.
--

redpile.behavior('push_solid', {'PUSH', 'PULL'}, function(node, messages)
    if messages.count > 0 then
        local message = messages:first()
        node:move(message.value)
    end
end)

redpile.behavior('push_breakable', {'PUSH'}, function(node, messages)
    if messages.count > 0 then
        node:remove()
    end
end)

redpile.behavior('push_piston', {'PUSH', 'PULL'}, function(node, messages)
    if messages.count > 0 and node.state == RETRACTED then
        local message = messages:first()
        node:move(message.value)
    end
end)

redpile.behavior('power_wire', {'POWER'}, function(node, messages)
    local covered = node:adjacent(UP).type ~= 'AIR'
    local power_msg = messages:max()
    node.power = power_msg and power_msg.value or 0

    if node.power == 0 then
        return
    end

    node:adjacent_each(DOWN, function(found)
        if found.type == 'CONDUCTOR' and
           has_lower_power(found, messages, node.power)
       then
           found:send('POWER', 0, node.power)
       end
    end)

    if node.power == 1 then
        return
    end

    local wire_power = node.power - 1

    node:adjacent_each(NORTH, SOUTH, EAST, WEST, function(found, dir)
        if found.type == 'AIR' then
            found = found:adjacent(DOWN)
            if found.type ~= 'WIRE' then
                return
            end
        elseif found.type == 'CONDUCTOR' then
            if node:adjacent(redpile.direction_left(dir)).type ~= 'WIRE' and
               node:adjacent(redpile.direction_right(dir)).type ~= 'WIRE' and
               has_lower_power(found, messages, wire_power)
           then
               found:send('POWER', 0, wire_power)
           end

           if covered then
               return
           end

           found = found:adjacent(UP)
           if found.type ~= 'WIRE' then
               return
           end
        end

        if has_lower_power(found, messages, wire_power) then
            found:send('POWER', 0, wire_power)
        end
    end)
end)

redpile.behavior('power_conductor', {'POWER'}, function(node, messages)
    node.power = value_or_zero(messages:max())
    local max_powerd = node.power == MAX_POWER

    node:adjacent_each(function(found)
        if found.type ~= 'CONDUCTOR' and (max_powerd or found.type ~= 'WIRE') then
            if has_lower_power(found, messages, node.power) then
                found:send('POWER', 0, node.power)
            end
        end
    end)
end)

redpile.behavior('power_torch', {'POWER'}, function(node, messages)
    local new_power = value_or_zero(messages:source(BEHIND))
    if new_power > 0 then
        node.power = 0
        return
    end
    node.power = MAX_POWER

    local behind = node:adjacent(BEHIND)
    node:adjacent_each(NORTH, SOUTH, EAST, WEST, DOWN, function(found)
        if found.location ~= behind.location then
            found:send('POWER', 1, MAX_POWER)
        end
    end)

    node:adjacent_each(UP, function(found)
        if found.type == 'CONDUCTOR' then
            found:send('POWER', 1, MAX_POWER)
        end
    end)
end)

redpile.behavior('power_piston', {'POWER'}, function(node, messages)
    local first = node:adjacent(FORWARDS)
    local second = first:adjacent(node.direction)
    local new_power = value_or_zero(messages:max())

    if new_power == 0 then
        if first.type == 'AIR' and second.type ~= 'AIR' and node.power > 0 then
            node.state = RETRACTING
        else
            node.state = RETRACTED
        end
    else
        if second.type == 'AIR' and first.type ~= 'AIR' and node.power == 0 then
            node.state = EXTENDING
        else
            node.state = EXTENDED
        end
    end

    node.power = new_power

    if node.state == EXTENDING then
        first:send('PUSH', 1, node.direction)
    elseif node.state == RETRACTING then
        second:send('PULL', 1, redpile.direction_invert(node.direction))
    end
end)

redpile.behavior('power_repeater', {'POWER'}, function(node, messages)
    node.power = value_or_zero(messages:source(BEHIND))
    if node.power ~= 0 and
       messages:source(RIGHT) == nil and
       messages:source(LEFT) == nil
   then
       node:adjacent(FORWARDS):send('POWER', node.state + 1, MAX_POWER)
   end
end)

redpile.behavior('power_comparator', {'POWER'}, function(node, messages)
    node.power = value_or_zero(messages:source(BEHIND))
    if node.power == 0 then
        return
    end

    local side_power = math.max(
        value_or_zero(messages:source(LEFT)),
        value_or_zero(messages:source(RIGHT))
    )

    local change = node.power
    if node.state > 0 then
        change = change - side_power
    end

    local new_power = (node.power > side_power) and change or 0
    if new_power ~= 0 then
        node:adjacent(FORWARDS):send('POWER', 1, new_power)
    end
end)

redpile.behavior('power_switch', {}, function(node, messages)
    if node.state == 0 then
        node.power = 0
        return
    end

    node.power = MAX_POWER
    local behind = node:adjacent(BEHIND)
    node:adjacent_each(function(found)
        if found.type ~= 'CONDUCTOR' or found.location == behind.location then
            found:send('POWER', 0, MAX_POWER)
        end
    end)
end)

redpile.behavior('power_echo', {'POWER'}, function(node, messages)
    node.power = value_or_zero(messages:max())
    if node.power > 0 and node.message then
        node:data(node.message)
    end
end)

-- Types are defined using the `redpile.type` function which takes:
--
-- NAME <String>
-- The name used to reference this type.  Should be upper case.
--
-- FIELDS <Table>
-- Table of fields to store on this node.  Declare in the format:
--   {FIELD_NAME} = {FIELD_TYPE}
-- Available field types are:
--   FIELD_INTEGER
--   FIELD_DIRECTION
--   FIELD_STRING
--
-- BEHAVIORS <Table>
-- Table containing names of behaviors for this type that will be run in the
-- order listed.
--

redpile.type(
    'AIR',
    {},
    {}
)

redpile.type(
    'INSULATOR',
    {},
    {'push_solid'}
)

redpile.type(
    'WIRE',
    {power = FIELD_INTEGER},
    {'push_breakable', 'power_wire'}
)

redpile.type(
    'CONDUCTOR',
    {power = FIELD_INTEGER},
    {'push_solid', 'power_conductor'}
)

redpile.type(
    'TORCH',
    {power = FIELD_INTEGER, direction = FIELD_DIRECTION},
    {'push_breakable', 'power_torch'}
)

redpile.type(
    'PISTON',
    {power = FIELD_INTEGER, direction = FIELD_DIRECTION, state = FIELD_INTEGER},
    {'push_piston', 'power_piston'}
)

redpile.type(
    'REPEATER',
    {power = FIELD_INTEGER, direction = FIELD_DIRECTION, state = FIELD_INTEGER},
    {'push_breakable', 'power_repeater'}
)

redpile.type(
    'COMPARATOR',
    {power = FIELD_INTEGER, direction =  FIELD_DIRECTION, state = FIELD_INTEGER},
    {'push_breakable', 'power_comparator'}
)

redpile.type(
    'SWITCH',
    {power = FIELD_INTEGER, direction = FIELD_DIRECTION, state = FIELD_INTEGER},
    {'push_breakable', 'power_switch'}
)

redpile.type(
    'ECHO',
    {power = FIELD_INTEGER, message = FIELD_STRING},
    {'power_echo'}
)

